Add docs for writing a new widget (#7075)
plone/volto: docs/source/development/widget.md

      />
```

> You'll need to define a new KeywordsIndex in `portal_catalog`, in a {file}`catalog.xml` GenericSetup file.

Reviewer: ```suggestion
You'll need to define a new `KeywordsIndex` in `portal_catalog`, in a {file}`catalog.xml` GenericSetup file.
```

## Write a new widget

A custom widget is often used in Volto forms.
To configure a new Volto widget follow these steps.

> 1. Add this import in the first line of {file}`index.js` file of the addon to import the component that is the actual widget.
> 
> ```tsx
> import { ToggleWidget } from './ToggleWidget';
> ```

Reviewer: Code blocks must be indented to the same level as the list item's first character so that they are within the context of the list item. Because of this, I strongly recommend always using 4 spaces for indenting list items, since that is a common code editor default.

Additionally, enumerated items may use `1.` for all items, letting MyST automatically handle the increasing enumeration for you, especially if you need to insert or remove an item from the list.

I've only added one suggestion for the first enumerated list item, but you should handle the others in the same fashion.

Also "add-on" is the proper spelling when used in narrative text, but may be "addon" in code and other inline literals.

````suggestion
1.  Add this import in the first line of {file}`index.js` file of the add-on to import the component that is the actual widget.

    ```tsx
    import { ToggleWidget } from './ToggleWidget';
    ```
````

4. Using the component shadowing tehnique, inside a new directory of your addon: {file}`src/customizations/components/manage/Sidebar/` create a file {file}`Sidebar.jsx` that shadows the Sidebar component inside the core Volto corresponding file.
Copy all the file contents of original {file}`Sidebar.jsx` to the new {file}`Sidebar.jsx`.

5. Inside the new {file}`Sidebar.jsx`, add this import at the top:

```jsx
import { Form } from '@plone/volto/components/manage/Form';
```

Below the imports, write this example custom form which has one field of type `toggle_widget` (our widget):

```jsx
const MyForm = () => {
  return (
    <Form
      schema={{
        fieldsets: [
          {
            id: 'default',
            title: 'Title',
            fields: ['toggle'],
          },
        ],
        properties: {
          toggle: {
            title: 'My toggle',
            widget: 'toggle_widget',
            default: 'On',
          },
        },
        required: [],
      }}
    />
  );
};
```

To see the widget inside the `Sidebar` on the right of Volto UI, add this line:

```jsx
<MyForm />
```

inside the `(React.)Fragment` tag that is rendered.

> Now, after restarting Volto, you can see the button the widget renders, click on it, and its text changes on each click.

Reviewer: Should this be step 6?

    ```jsx
    import { ToggleWidget } from './ToggleWidget';
    ```

1.  Add the following line before returning from the `applyConfig` function inside the file {file}`index.js` in the add-on to register the widget add-on component.

    ```jsx
    config.widgets.widget.toggle_widget = ToggleWidget;
    ```

1.  Next, create the widget component file {file}`ToggleWidget.jsx` with the following content.

    ```jsx
    import { Button } from "semantic-ui-react";
    export function ToggleWidget({ id, value, onChange }) {
      return (
        <Button
          onClick={(e) => {
            e.preventDefault();
            onChange(id, value === "On" ? "Off" : "On");
          }}
        >
          Toggle me: {value}
        </Button>
      );
    }

    export default ToggleWidget;
    ```

1.  Install `semantic-ui-react` as a dependency of your add-on to make it work.

    ```shell
    npm i semantic-ui-react
    ```

1.  Using the component shadowing technique, create a directory in your add-on {file}`src/customizations/components/manage/Sidebar/`.

>      Copy the file in Volto core {file}`components/manage/Sidebar/Sidebar.jsx` to the directory you just created.
>      This file shadows the `Sidebar` component inside the corresponding core Volto file.

Reviewer: My bad. One too many leading spaces for subsequent sentences.

```suggestion
    Copy the file in Volto core {file}`components/manage/Sidebar/Sidebar.jsx` to the directory you just created.
    This file shadows the `Sidebar` component inside the corresponding core Volto file.
```

1.  Inside the new {file}`src/customizations/components/manage/Sidebar/Sidebar.jsx`, insert the following import at the top.

    ```jsx
    import { Form } from '@plone/volto/components/manage/Form';
    ```

1.  Below the imports, write the following example custom form, which has one field of type `toggle_widget` to serve as our widget.

    ```jsx
    const MyForm = () => {
      return (
        <Form
          schema={{
            fieldsets: [
              {
                id: "default",
                title: "Title",
                fields: ["toggle"],
              },
            ],
            properties: {
              toggle: {
                title: "My toggle",
                widget: "toggle_widget",
                default: "On",
              },
            },
            required: [],
          }}
        />
      );
    };
    ```

1.  To see the widget inside the `Sidebar` component on the right side of the Volto UI, add the following line inside the `(React.)Fragment` tag that is rendered.

    ```jsx
    <MyForm />
    ```

> 1.  Finally, restart Volto by pressing Ctrl-C inside the shell instance which runs it, then restart it with the following command.

Reviewer: We have references for MyST markup, listed in order of which to check. Plone's reference is limited to the most frequently used markup, then MyST which has most roles and directives, and ultimately Sphinx. Unfortunately this is one of those that is only documented in the grandparent in Sphinx's documentation.

- https://6.docs.plone.org/contributing/documentation/myst-reference.html
- https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html#roles-an-in-line-extension-point
- https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-kbd

```suggestion
1.  Finally, restart Volto by pressing {kbd}`Ctrl-C` inside the active frontend shell session, then restart it with the following command.
```

    ```shell

>     make frontend-start

Reviewer: Did this command actually work? I went from memory, as I don't have an actual project to test it on. It could be `make start` per https://6.docs.plone.org/volto/development/add-ons/create-an-add-on-18.html#start-plone-development-frontend.

Author: > Did this command actually work? I went from memory, as I don't have an actual project to test it on. It could be `make start` per [6.docs.plone.org/volto/development/add-ons/create-an-add-on-18.html#start-plone-development-frontend](https://6.docs.plone.org/volto/development/add-ons/create-an-add-on-18.html#start-plone-development-frontend).

It works in the root of frontend addon projects created with Cookieplone. Please let me know if I can click the Resolve conversation button.

Reviewer: Thanks for confirming. Resolved.

    ```

Now you can see the button that the widget renders.
Click the button to toggle the text.
You can position the `MyForm` component anywhere where it is visible and the user can interact with it.

> You can check other widgets included in Volto core under the widgets directory for more inspiration and possibilities of enhancing your widgets.

Reviewer: Let's make this a separate paragraph, as it leads the reader to a new challenge.

```suggestion

You can experiment with other widgets included in Volto core under the {file}`Widgets` directory for more inspiration and possibilities of enhancing your widgets.
```

## Write a new widget

> A custom widget is often used in Volto forms.

Reviewer 2: Please add reasons why I would want to write a custom widget.

> To configure a new Volto widget follow these steps.

Reviewer 2: a short list of steps before going into details would be nice.

Reviewer: I don't understand why it would be nice and not redundant.

> 1.  Add the following import as the first line in the file {file}`index.js` in the add-on to import the component of the actual widget.

Reviewer 2: maybe first create component, then register it.

    ```jsx
    import { ToggleWidget } from './ToggleWidget';
    ```

1.  Add the following line before returning from the `applyConfig` function inside the file {file}`index.js` in the add-on to register the widget add-on component.

    ```jsx

>     config.widgets.widget.toggle_widget = ToggleWidget;

Reviewer 2: can we add please the registration for boolean fields.

    ```

1.  Next, create the widget component file {file}`ToggleWidget.jsx` with the following content.

    ```jsx
    import { Button } from "semantic-ui-react";
    export function ToggleWidget({ id, value, onChange }) {
      return (

>         <Button

Reviewer 2: why a button? why not a toggle?

          onClick={(e) => {
            e.preventDefault();
            onChange(id, value === "On" ? "Off" : "On");
          }}
        >
          Toggle me: {value}
        </Button>
      );
    }

    export default ToggleWidget;
    ```

> 1.  Install `semantic-ui-react` as a dependency of your add-on to make it work.

Reviewer 2: `semantic-ui-react` is already included in a Volto project, if you do not explicitly exclude it because you want to avoid it.

Reviewer: Good point. We should not encourage the use of Semantic UI, now that it is deprecated. See https://6.docs.plone.org/volto/release-notes/index.html#id29 and an example at https://6.docs.plone.org/volto/theming/index.html.

    ```shell
    npm i semantic-ui-react
    ```

> 1.  Using the component shadowing technique, create a directory in your add-on {file}`src/customizations/components/manage/Sidebar/`.

Reviewer 2: I would prefer an example with a standard `Form` from '@plone/volto/components/manage/Form/Form';

    This file shadows the `Sidebar` component inside the corresponding core Volto file.

1.  Inside the new {file}`src/customizations/components/manage/Sidebar/Sidebar.jsx`, insert the following import at the top.

    ```jsx
    import { Form } from '@plone/volto/components/manage/Form';
    ```

1.  Below the imports, write the following example custom form, which has one field of type `toggle_widget` to serve as our widget.

    ```jsx
    const MyForm = () => {
      return (
        <Form
          schema={{
            fieldsets: [
              {
                id: "default",
                title: "Title",
                fields: ["toggle"],
              },
            ],
            properties: {
              toggle: {
                title: "My toggle",
                widget: "toggle_widget",
                default: "On",
              },
            },
            required: [],
          }}
        />
      );
    };
    ```

1.  To see the widget inside the `Sidebar` component on the right side of the Volto UI, add the following line inside the `(React.)Fragment` tag that is rendered.

    ```jsx
    <MyForm />
    ```

> 1.  Finally, restart Volto by pressing {kbd}`Ctrl-C` inside the active frontend shell session, then restart it with the following command.

Reviewer 2: I think "restart Volto" is enough. We know how to do it.

Reviewer: Let's help first-timers. It's OK to include the commands.

## Write a new widget

A custom widget is often used in Volto forms.
To configure a new Volto widget follow these steps.

> 1.  Add this import in the first line of {file}`index.js` file of the add-on to import the component that is the actual widget.

Reviewer: ```suggestion
1.  Add the following import as the first line in the file {file}`index.js` in the add-on to import the component of the actual widget.
```

    ```jsx
    import { ToggleWidget } from './ToggleWidget';
    ```

> 1.  Add this line before returning from `applyConfig` function inside the {file}`index.js` of the add-on to register the add-on component (widget).

Reviewer: ```suggestion
1.  Add the following line before returning from the `applyConfig` function inside the file {file}`index.js` in the add-on to register the widget add-on component.
```

    ```jsx
    config.widgets.widget.toggle_widget = ToggleWidget;
    ```

> 1.  Next, create the widget component file {file}`ToggleWidget.jsx` with these contents:

Reviewer: ```suggestion
1.  Next, create the widget component file {file}`ToggleWidget.jsx` with the following content.
```

    ```jsx

>     import { Button } from 'semantic-ui-react';
> 
>     export function ToggleWidget({ id, value, onChange }) {
>         return (
>             <Button
>                 onClick={(e) => {
>                     e.preventDefault();
>                     onChange(id, value === 'On' ? 'Off' : 'On');
>                 }}
>             >
>                 Toggle me: {value}
>             </Button>
>         );
>     }
> 
>     export default ToggleWidget;

Reviewer: I ran this through prettier. Code blocks should follow the project's formatting guidance.

```suggestion
    import { Button } from "semantic-ui-react";
    export function ToggleWidget({ id, value, onChange }) {
      return (
        <Button
          onClick={(e) => {
            e.preventDefault();
            onChange(id, value === "On" ? "Off" : "On");
          }}
        >
          Toggle me: {value}
        </Button>
      );
    }

    export default ToggleWidget;
```

    ```

>     You need to install `semantic-ui-react` as a dependency of your add-on before it will work.
>     You only need to `npm i semantic-ui-react`.

Reviewer: ```suggestion
1.  Install `semantic-ui-react` as a dependency of your add-on to make it work.

    ```shell
    npm i semantic-ui-react
    ```
```

1.  Using the component shadowing tehnique, inside a new directory of your add-on: {file}`src/customizations/components/manage/Sidebar/` create a file {file}`Sidebar.jsx` that shadows the Sidebar component inside the core Volto corresponding file.

> Copy all the file contents of original {file}`Sidebar.jsx` to the new {file}`Sidebar.jsx`.

Reviewer: ```suggestion
```

> 1.  Inside the new {file}`Sidebar.jsx`, add this import at the top:

Reviewer: ```suggestion
1.  Inside the new {file}`src/customizations/components/manage/Sidebar/Sidebar.jsx`, insert the following import at the top.
```

    ```jsx
    import { Form } from '@plone/volto/components/manage/Form';
    ```

> 1.  Below the imports, write this example custom form which has one field of type `toggle_widget` (our widget):

Reviewer: ```suggestion
1.  Below the imports, write the following example custom form, which has one field of type `toggle_widget` to serve as our widget.
```

    ```jsx
    const MyForm = () => {
        return (
            <Form
                schema={{
                    fieldsets: [
                        {
                            id: 'default',
                            title: 'Title',
                            fields: ['toggle'],
                        },
                    ],
                    properties: {
                        toggle: {
                            title: 'My toggle',
                            widget: 'toggle_widget',
                            default: 'On',
                        },
                    },
                    required: [],
                }}
            />
        );
    };
    ```

1.  To see the widget inside the `Sidebar` on the right of Volto UI, add this line:

    ```jsx
    <MyForm />
    ```

>     inside the `(React.)Fragment` tag that is rendered.

Author: @stevepiercy I've discovered that in a full project template project of Cookieplone, the make frontend-start works well. But in an add-on project template project of Cookieplone, there is just make start and it works well. Is it OK to put both commands in there?

>     inside the `(React.)Fragment` tag that is rendered.

Reviewer: Yes, please. And you can use sphinx-design with a tabbed interface to present the two installation methods side by side, with one as the default. See https://6.docs.plone.org/contributing/documentation/myst-reference.html#tabs

>     inside the `(React.)Fragment` tag that is rendered.

Author: @stevepiercy Done.

1.  Finally, restart Volto by pressing {kbd}`Ctrl-C` inside the active frontend shell session, then restart it with the following command.

>     ````{tab-set}
>       ```{tab-item} Frontend add-on Cookieplone project
>         ```shell
>         make start
>         ```
>       ```
>       ```{tab-item} Fullstack Cookieplone project
>         ```shell
>         make frontend-start
>         ```
>       ```
>     ````

Reviewer: That's creative syntax, and I'm surprised it works. Let's stick with supported MyST syntax, and nest by increasing the count of backticks, in case it becomes strict.

```suggestion
    `````{tab-set}
    ````{tab-item} Frontend add-on Cookieplone project
    ```shell
    make start
    ```
    ````
    ````{tab-item} Fullstack Cookieplone project
    ```shell
    make frontend-start
    ```
    ````
    `````
```
